src/Razor/Microsoft.AspNetCore.Razor.Language/test/IntegrationTests/ComponentCodeGenerationTestBase.cs

```diff
@@ -3152,6 +3152,391 @@ public class MyComponent<TItem> : ComponentBase
+            // Arrange
+            AdditionalSyntaxTrees.Add(Parse(@"
+using Microsoft.AspNetCore.Components;
+
+namespace Test
+{
+    [CascadingTypeParameter(nameof(TItem))]
+    public class Grid<TItem> : ComponentBase
+    {
+        [Parameter] public System.Collections.Generic.IEnumerable<TItem> Items { get; set; }
+        [Parameter] public RenderFragment ChildContent { get; set; }
+    }
+
+    public class Column<TItem> : ComponentBase
+    {
+    }
+}
+"));
+
+            // Act
+            var generated = CompileToCSharp(@"
+<Grid TItem=""DateTime"" Items=""@(Array.Empty<DateTime>())""><Column /><Column /></Grid>");
+
+            // Assert
+            AssertDocumentNodeMatchesBaseline(generated.CodeDocument);
+            AssertCSharpDocumentMatchesBaseline(generated.CodeDocument);
+            CompileToAssembly(generated);
+        }
+
+        [Fact]
+        public void ChildComponent_Generic_Explicit_OverrideCascade()
+        {
+            // Arrange
+            AdditionalSyntaxTrees.Add(Parse(@"
+using Microsoft.AspNetCore.Components;
+
+namespace Test
+{
+    [CascadingTypeParameter(nameof(TItem))]
+    public class Grid<TItem> : ComponentBase
+    {
+        [Parameter] public System.Collections.Generic.IEnumerable<TItem> Items { get; set; }
+        [Parameter] public RenderFragment ChildContent { get; set; }
+    }
+
+    public class Column<TItem> : ComponentBase
+    {
+    }
+}
+"));
+
+            // Act
+            var generated = CompileToCSharp(@"
+<Grid TItem=""DateTime"" Items=""@(Array.Empty<DateTime>())""><Column TItem=""System.TimeZoneInfo"" /></Grid>");
+
+            // Assert
+            AssertDocumentNodeMatchesBaseline(generated.CodeDocument);
+            AssertCSharpDocumentMatchesBaseline(generated.CodeDocument);
+            CompileToAssembly(generated);
+        }
+
+        [Fact]
+        public void ChildComponent_Generic_Explicit_NotCascaded()
+        {
+            // The point of this test is to show that, without [CascadingTypeParameter], we don't cascade
+
+            // Arrange
+            AdditionalSyntaxTrees.Add(Parse(@"
+using Microsoft.AspNetCore.Components;
+
+namespace Test
+{
+    public class Grid<TItem> : ComponentBase
+    {
+        [Parameter] public System.Collections.Generic.IEnumerable<TItem> Items { get; set; }
+        [Parameter] public RenderFragment ChildContent { get; set; }
+    }
+
+    public class Column<TItem> : ComponentBase
+    {
+    }
+}
+"));
+
+            // Act
+            var generated = CompileToCSharp(@"
+<Grid TItem=""DateTime"" Items=""@(Array.Empty<DateTime>())""><Column /></Grid>");
+
+            // Assert
+            AssertDocumentNodeMatchesBaseline(generated.CodeDocument);
+            AssertCSharpDocumentMatchesBaseline(generated.CodeDocument);
+
+            var diagnostic = Assert.Single(generated.Diagnostics);
+            Assert.Same(ComponentDiagnosticFactory.GenericComponentTypeInferenceUnderspecified.Id, diagnostic.Id);
```

Comment:
It might be good to update this diagnostic message to include the list of considered and rejected candidates. Right now, if type inference fails, there is not a good way to know why.

Reply:
It would be nice if we had more useful information to give, but I don't see that we do. Candidates are matched on name, so it's not useful to the developer to talk about things with unrelated names - that would be more confusing. Likewise, talking about all the ancestor components would be confusing too because there's no reason to think they should be involved in cascading.

The likely case where someone hits this doesn't involve cascading at all - it's the existing case where someone simply fails to supply a sufficiently covering set of parameters.

If we spot any common patterns in support requests we can certainly try to work out whether they are things we could detect programmatically.

Reply:
I was thinking we could list the ancestors in the tree that offer cascading parameter values and their names. That at least gives a hint of what was considered and can help troubleshoot when a parameter is not being inferenced.

The way I see this, is we go through the process of determining if the parameters a component supplies can be applied to a component that receives them for all the ancestors in the tree. At that point, we could record the reason why we think some component exposing cascading values is not a good match for the types that require inference and to display that information along the diagnostic error in case we aren't able to infer the parameters for the child component. Many compilers do this type of thing, as well as some parts of our runtime, like routing.

It helps people solve the issue by themselves in many cases and limits the number of issues we receive because developers don't understand the rules and can't figure out why it's not working.

Something along the lines of:
The type for TItem was not specified and could not be inferred from its ancestors. The following candidates were considered:
  * Parent cascades TKey, TValue parameters, but they were rejected because X.
  * Grandparent cascades TContext but was rejected because Y.
Either update a component on the ancestor chain to cascade the parameter TItem or specify the attribute explicitly.

Not exactly what I wrote above, but something along the lines.

I know this is more work (believe me, I implemented this for routing) but it greatly increases the "explainability" of the system. If you still don't think it's worth it, at least, let's file an issue to track this improvement and schedule on a future milestone if we receive feedback about this.

Reply:
It's an honourable thing to try to make better error messages, but I think something along the following lines risks causing a lot more confusion than would exist without it:

```
The type for TItem was not specified and could not be inferred from its ancestors. The following candidates were considered:

Parent cascades TKey, TValue parameters, but they were rejected because X.
Grandparent cascades TContext but was rejected because Y.
Either update a component on the ancestor chain to cascade the parameter TItem or specify the attribute explicitly.
```

Talking about cascading generic types in the error makes the developer think that cascading is somehow part of either the problem or the solution. But in the typical case, cascading is completely unrelated to the situation. The most common case is that the component isn't meant to receive the type via a cascade at all, but is meant to receive it via one of its own parameters.

Certainly if we get some patterns of feedback saying that it's hard to understand *and* we believe we can distinguish those cases where cascading is intended as the solution, then we can consider making an error message that leads the developer to think about that. But let's be careful not to mislead developers in the more common case of just using a normally type-inferred component.